Add capability to redirect to Python logger

jl-wynen opened the pull request:
**Description of work.**

Defines a new `Poco::Channel` which sends messages to a logger from Python's `logging` framework and `log_to_python` to help set up the logger.

The Python logger has fewer levels than Poco's. It is possible to define new levels in Python but those might clash with user defined levels. Given that the new mechanism is meant to be used in combination with other packages that might log, defining new levels could cause problems. This PR, therefore,  maps several Poco levels onto the same Python level.

It is unclear to me how this should interact with the rest of Mantid's logging config. This PR does not break any existing behaviour and leaves it up to the user to configure everything as they need it.

**To test:**
```.py
import logging
from mantid.utils.logging import log_to_python
from mantid.kernel import logger

logging.basicConfig(format='PYTHON-LOGGER:: <%(name)s> %(levelname)s : %(message)s',
                    level=logging.INFO)
log = logging.getLogger()

log_to_python()

logger.information('[[info]]')
logger.warning('[[warning]]')
logger.error('[[error]]')
logger.fatal('[[fatal]]')
```
should print
```
PYTHON-LOGGER:: <Mantid> INFO : [[info]]
PYTHON-LOGGER:: <Mantid> WARNING : [[warning]]
PYTHON-LOGGER:: <Mantid> ERROR : [[error]]
PYTHON-LOGGER:: <Mantid> CRITICAL : [[fatal]]
```

See also new `loggingTest.test_log_to_python`.

Fixes #33215

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

jl-wynen committed:
Add PythonLoggingChannel

jl-wynen committed:
Remove explicit destructor of PythonLoggingChannel

jl-wynen committed:
Store Python log levels in enum

jl-wynen committed:
Register PytonLoggingChannel with Poco

jl-wynen committed:
Looks like we do not need to acquire the GIL ourselves

jl-wynen committed:
Add comment to PythonLoggingChannel.h

jl-wynen committed:
Acquire GIL before logging

jl-wynen committed:
Only destroy Python logger then interpreter is running

jl-wynen committed:
Add log_to_python

jl-wynen committed:
Add test_log_to_python

jl-wynen committed:
Add python logging section to docs

jl-wynen committed:
Add release note on logging to Python

jl-wynen committed:
Only access python logger if interpreter is initialized

jl-wynen commented:
Comments in the issue by @peterfpeterson 
> and am not sure that log_to_python() is a necessary function rather than instructing users on how to modify their properties file. 

I added it as a convenience, especially because it depends on the order in which you set the keys. But it is not a requirement to make the logger usable.

> On linux, there is a system level properties file that could set the logging channel for users on that machine as a whole.

Using a system or even user wide setting would mean that it is used by both the GUI and Python package. This seems like a bad idea to me.

jl-wynen committed:
Document PythonLoggingChannel in PropertiesFile.rst

peterfpeterson commented:
Bootstrapping the logging system is a pain. Current workbench and friends can't start logging until configservice and logging are stood up anyhow. Since python is integral to mantid working, I don't think it is a huge issue if you have to have python up to log.

jl-wynen committed:
Merge branch 'main' into python-logging

jl-wynen committed:
Move logging release note to 6.4.0

peterfpeterson commented on Framework/PythonInterface/core/src/PythonLoggingChannel.cpp:
```diff
@@ -0,0 +1,80 @@
+// after the Python interpreter was shut down.
+PythonLoggingChannel::~PythonLoggingChannel() {
+  if (Py_IsInitialized()) {
+    Mantid::PythonInterface::GlobalInterpreterLock gil;
```
I may be reading this wrong, but I think `release` should happen if python is initialized and setting it to `nullptr` should happen if python isn't initialized.

jl-wynen replied:
`m_pyLogger = nullptr;` makes `std::unique_ptr` delete the contained object which calls its destructor which in turn needs to access Python.
The `release` hack allows us to only free the memory allocated by us without calling into Python via `*m_pyLogger`'s destructor.

peterfpeterson commented on Framework/PythonInterface/test/python/mantid/utils/loggingTest.py:
```diff
@@ -7,14 +7,34 @@
+
+
+@contextmanager
+def backup_config():
```
It feels like there is a better place to put this. I've written this same functionality myself a bunch of times which isn't very DRY.

jl-wynen replied:
How about
```.py
config = ConfigService.Instance()
with config.temporary():
    config[key] = value
    # do stuff
# config gets reset here
```
But I don't know of the larger implications here. Would backing up and restoring the entire config re-init important components and cause problems?

Or are you referring to the logging config specifically?

peterfpeterson replied:
I meant as a utility for testing, not specific to logging. If you can think of a good place to put this so we can advertise it to developers I'd be very appreciative. However, it is outside the scope of this PR so I won't block merging based on this.

jl-wynen replied:
Looks like `Framework/PythonInterface/test/testhelpers` is a good place. Could be either in `__init__.py` or in a new separate module.

peterfpeterson commented on dev-docs/source/Logging.rst:
```diff
@@ -126,6 +126,26 @@ In general:
   from mantid.kernel import logger
   logger.warning('this is a custom warning')
 
+Using :code:`logging` in Python
```
Can you add a link to/form `docs/source/api/python/mantid/kernel/Logger.rst` in this as well. That documentation is auto-generated and not nearly as useful.

peterfpeterson replied:
The link here should be
```rst
:class:`mantid.kernel.Logger`
```

jl-wynen replied:
Ok. I thought you meant to link the other way round.

peterfpeterson commented on docs/source/release/v6.4.0/framework.rst:
```diff
@@ -21,6 +21,8 @@ Data Objects
 Python
 ------
 
+- Added possibility to forward log messages to Python, see `mantid.utils.logging.log_to_python`.
```
Please make this a link to `dev-docs/source/Logging.rst`

jl-wynen replied:
How do you link from the user to the dev docs? Is there a nicer way than using URLs to the published online documentation?

peterfpeterson replied:
It should be
```
:ref:`mantid.utils.logging.log_to_python`
```

jl-wynen replied:
Ah, ok.
But I was also asking about your other comment where I need to link to  a file / section.

SimonHeybrock reviewed (dismissed):
Nice, looks good to me, aside from Pete's comments. 👍

jl-wynen committed:
Use link in release note

jl-wynen committed:
Link to mantid.kernel.Logger

jl-wynen committed:
Add testhelpers.temporary_config

jl-wynen committed:
Use temporary_config for logging tests

jl-wynen committed:
Merge branch 'mantidproject:main' into python-logging

jl-wynen committed:
Store Python log levels as constexpr int

MSVC did not like the previous solution using a typed enum. It reported a syntax error because of a missing '}'.

jl-wynen commented:
I don't know what is wrong with the failing test. It looks like a problem with QT and I don't see how this PR caused it. Can someone help?

SimonHeybrock commented:
A similar issue was reported in #unreliable-tests a while back, I am trying to restart your build for now.
